improve hydro stability

xuchongang opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This pull request improves the numerical stability in the situation of extreme dry conditions
### Description:
Hydro uses the transpiration flux from last time step. For numerical stability, this pull request make sure that transpiration flux is less than 80% of total water availability. If it is larger than that, it will adjust the transpiration flux and record the excessive demand to the plant water storage pool, which will be dumped later in HLM.This should not affect the science of hydrodynamics as this basically delays the water limitation by a few steps with a benefit of numerical stability. 

### Collaborators:
@bchristo @rgknox 

### Expectation of Answer Changes:
This will not affect most of simulation and should only affect simulations with very dry soil.


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [x] I have updated the in-code documentation .AND. (the [technical note](https://github.com/NGEET/fates-docs) .OR. the wiki) accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [x] FATES PASS/FAIL regression tests were run
- [x] If answers were expected to change, evaluation was performed and provided


### Test Results:
<!--- Non-trivial changes require the PASS/FAIL regression tests. -->
<!--- If changes to code are NOT expected to change answers, tests must -->
<!--- be run against a baseline. -->

CTSM (or) E3SM (specify which) test hash-tag:

CTSM (or) E3SM (specify which) baseline hash-tag:

FATES baseline hash-tag:

Test Output:

<!--- paste in test results here -->


<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

xuchongang committed:
put a transpiration flux check to increase numerical stability

xuchongang committed:
add a transpiration error term

rgknox commented:
@xuchongang , I think we might need to subtrace the trans_err term to the plant stored water output boundary condition, here:

https://github.com/NGEET/fates/blob/aa290b3b8ad95dc86214cb3ac0ff7772128514cf/biogeophys/FatesPlantHydraulicsMod.F90#L1810-L1812

That way when E3SM and CTSM do their calculations on mass conservation, it will include this water we needed to create.

xuchongang commented:
> @xuchongang , I think we might need to subtract the trans_err term to the plant stored water output boundary condition, here:
> 
> https://github.com/NGEET/fates/blob/aa290b3b8ad95dc86214cb3ac0ff7772128514cf/biogeophys/FatesPlantHydraulicsMod.F90#L1810-L1812
> 
> That way when E3SM and CTSM do their calculations on mass conservation, it will include this water we needed to create.
> 
> Thoughts?

@rgknox , you are correct. I forget to copy that over as I am testing on the ELM, which is not the most recent compared to master. Will do!

xuchongang committed:
add the transpiration error term for mass balance check

mpaiao committed:
Update FatesPlantHydraulicsMod.F90

Fix a few typos in FatesPlantHydraulicsMod (no change in the code functionality).

mpaiao committed:
Unit conversion fix: FatesPlantHydraulicsMod.F90

Revised the transpiration error quantification. The previous version seemed to have the wrong units, and BalanceCheckMod.F90 was often failing because errh2o was several orders of magnitude larger than the tolerance. With these edits, all my test simulations seem to be going through, although I have not made extensive tests.

@xuchongang, could you please check and see if these changes make sense? Thanks!

mpaiao committed:
Update FatesPlantHydraulicsMod.F90

Fix minor typos to my previous commit. 

site_hydr was renamed to csite_hydr at some point, and my previous commit was using the old notation. I fixed that and also edited CalculateTotalAvailW to use csite and ccohort to be consistent with the other subroutines.

mpaiao committed:
Update FatesPlantHydraulicsMod.F90

Additional update of site_hydr name to csite_hydr.

rgknox committed:
merge resolution, unit evaluation

rgknox committed:
added unit fixes to hydro stability fix

rgknox committed:
Merge branch 'main' into xuchongang/master-hydro-stability

rgknox reviewed (approved):
@xuchongang and I worked through a unit fix, which was applied.  Looks good.
